src/Lecture2.hs

```diff
@@ -288,4 +364,23 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding = addExpr . extractConstant
+    where
+      extractConstant :: Expr -> ([Int], [Expr])
+      extractConstant (Lit i)   = ([i], [])
+      extractConstant (Var x)   = ([], [Var x])
+      extractConstant (Add x y) = (x1 ++ y1, x2 ++ y2)
+        where
+          (x1, x2) = extractConstant x
+          (y1, y2) = extractConstant y
+      addExpr :: ([Int], [Expr]) -> Expr
+      addExpr (c, e) = case (c, e) of
+        ([], []) -> error "invalid"
```

Comment:
Interesting! You can see that it's actually impossible to have both lists empty at the same time. So here you're handling and impossible situation 🙂 

Can you maybe change types a bit to avoid having this error? 😉

Reply:
@chshersh :-) I know you mean "making invalid states unrepresentable", need to think about this some more ...